lib/internal/modules/esm/loader.js

```diff
@@ -194,13 +218,63 @@ class ESMLoader {
         this.#loaders,
         FunctionPrototypeBind(loader, null), // [1]
       );
+      if (getFileSystem) ArrayPrototypePush(
+        this.#fileSystemBuilders,
+        FunctionPrototypeBind(getFileSystem, null), // [1]
+      );
     }
 
     // [1] ensure hook function is not bound to ESMLoader instance
 
+    this.buildFileSystem();
     this.preload();
   }
 
+  buildFileSystem() {
+    // Note: makes assumptions as to how chaining will work to demonstrate
+    // the capability; subject to change once chaining's API is finalized.
+    const fileSystemFactories = [...this.#fileSystemBuilders];
+
+    const defaultFileSystemFactory = fileSystemFactories[0];
+    let finalFileSystem =
+      defaultFileSystemFactory();
+
+    const asyncKeys = ArrayPrototypeFilter(
+      ObjectKeys(finalFileSystem),
+      (name) => !StringPrototypeEndsWith(name, 'Sync'),
+    );
+
+    for (let i = 1; i < fileSystemFactories.length; ++i) {
```

Comment:
tbh that's the first time I hear this - preincrementation are usually "preferred" in for loops since they don't require a temporary variable, although any decent compiler like v8 will surely treat them equally in this instance.

Do you have a particular case in mind where preincrementation leads to problems in a for loopexpression?